mbedtls/src/x509/certificate.rs

```diff
@@ -61,87 +85,142 @@ impl BERDecodable for Extension {
-            x509_crt_parse(&mut list.head.inner, pem.as_ptr(), pem.len()).into_result()?;
-        };
-        Ok(vec)
+    pub fn linked_cert(&self) -> LinkedCertificate<'_> {
+        LinkedCertificate { inner: &*self.inner }
     }
 }
 
 impl fmt::Debug for Certificate {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match crate::private::alloc_string_repeat(|buf, size| unsafe {
-            x509_crt_info(buf, size, b"\0".as_ptr() as *const _, &self.inner)
-        }) {
-            Err(_) => Err(fmt::Error),
-            Ok(s) => f.write_str(&s),
+        let mut count = 0;
+        for i in self.iter() {
+            f.write_fmt(format_args!("Certificate[{}]\n{:?}\n", count, i))?;
+            count = count + 1;
         }
+        Ok(())
     }
 }
 
 impl Clone for Certificate {
     fn clone(&self) -> Certificate {
-        let mut ret = Self::init();
+
         unsafe {
-            x509_crt_parse_der(&mut ret.inner, self.inner.raw.p, self.inner.raw.len)
-                .into_result()
-                .unwrap()
-        };
-        ret
+            let der = std::slice::from_raw_parts(self.inner.raw.p, self.inner.raw.len);
+            let mut cert = Certificate::from_der(&der).expect("Failed re-parsing existing DER");
+
+            let mut cur = self.inner.next;
+            while cur != ::core::ptr::null_mut() {
+                let der = std::slice::from_raw_parts((*cur).raw.p, (*cur).raw.len);
+                cert.push_back_der(&der).expect("Failed re-parsing existing DER");
+                cur = (*cur).next;
+            }
+
+            cert
+        }
     }
 }
 
-impl Deref for Certificate {
-    type Target = LinkedCertificate;
-    fn deref(&self) -> &LinkedCertificate {
-        unsafe { UnsafeFrom::from(&self.inner as *const _).unwrap() }
+impl Drop for Certificate {
+    fn drop(&mut self) {
+        // This frees ALL the linked certificates.
+        unsafe {
+            x509_crt_free(self.into());
+        }
     }
 }
 
-impl DerefMut for Certificate {
-    fn deref_mut(&mut self) -> &mut LinkedCertificate {
-        unsafe { UnsafeFrom::from(&mut self.inner as *mut _).unwrap() }
+pub struct Iter<'a> {
+    inner: Option<&'a x509_crt>
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = LinkedCertificate<'a>;
+
+    fn next(&mut self) -> Option<LinkedCertificate<'a>> {
+        match self.inner {
+            None => None,
+            Some(crt) => {
+                let ret = LinkedCertificate { inner: crt };
+                
+                if crt.next as *const _ != ::core::ptr::null() {
+                    self.inner = unsafe { Some(&*crt.next) };
+                } else {
+                    self.inner = None;
+                }
+                
+                Some(ret)
+            }
+        }
     }
 }
 
-#[repr(C)]
-pub struct LinkedCertificate {
-    inner: x509_crt,
+pub struct LinkedCertificate<'a> {
+    pub inner: &'a x509_crt
```

Comment:
Why are you changing this type?

Reply:
I do not want stack copies of the certificate as that makes possibly unsafe code.

With reference and lifetime I can make sure that LinkedCertificate only exists as long as parent certificate exists.
And since parent certificate makes sure nothing goes out of scope / gets delete - this ensures we are memory safe.

Reply:
There are not stack copies. We only ever create references to LinkedCertificate.

Reply:
This is how I create it now:
```
    pub fn linked_cert(&self) -> LinkedCertificate<'_> {
        LinkedCertificate { inner: &*self.inner }
    }
```

This is how it was created in the past:
```
impl Deref for Certificate {
    type Target = LinkedCertificate;
    fn deref(&self) -> &LinkedCertificate {
        unsafe { UnsafeFrom::from(&self.inner as *const _).unwrap() }
    }
}
```

So you are correct it only ever created references in the past.
But I'm reluctant to use the unsafe / UnsafeFrom. Also the Certificate may have a different structure.

Reply:
This was working correctly and reviewed. Please justify the change.

Reply:
It is simple to understand at first glance what this does:
```
    pub fn linked_cert(&self) -> LinkedCertificate<'_> {
        LinkedCertificate { inner: &*self.inner }
    }
```

Compared to old approach I linked above.

Let me know if I should revert.

Reply:
This is not a justification. If the change is not needed, then please revert yes.